File: chapter01/zeroMatrix/ZeroMatrix.java
public class ZeroMatrix {
    public static void main(String[] args) {
        int[][] matrix = {{1, 2, 3, 4, 0},
                {6, 0, 8, 9, 10},
                {11, 12, 13, 14, 15},
                {16, 0, 18, 19, 20},
                {21, 22, 23, 24, 25}};

        zeroMatrix(matrix);
    }

    public static void zeroMatrix(int[][] matrix) {
        if (matrix == null || matrix.length == 0) {
            return;
        }

        boolean[] rows = new boolean[matrix.length];

        for (int i = 0; i < matrix.length; i++) {
            boolean zeroRow = false;

            for (int j = 0; j < matrix[i].length; j++) {
                if (matrix[i][j] == 0 && !zeroRow) {
                    zeroRow = true;
                    rows[i] = true;
                }

                if (zeroRow) {
                    matrix[i][j] = 0;
                }
            }
        }

        System.out.println(printMatrix(matrix));
    }

    public static String printMatrix(int[][] matrix) {
        StringBuilder sb = new StringBuilder();
        sb.append("\n");

        for (int[] row : matrix) {
            for (int i : row) {
                sb.append(i).append(" ");
            }

            sb.append("\n");
        }

        return sb.toString();
    }
}

File: chapter01/zeroMatrix/ZeroMatrix.ts
function ZeroMatrix(matrix: number[][]): number[][] {
    let [m, n] = [matrix.length, matrix[0].length];
    let setZero = false;
    for (let row = 0; row < m; row++) {
        for (let col = 0; col < n; col++) {
            if (matrix[row][col] === 0) {
                matrix[0][col] = 0;

                if (row > 0) {
                    matrix[row][0] = 0;
                } else {
                    setZero = true;
                }
            }
        }
    }

    for (let row = 1; row < m; row++) {
        for (let col = 1; col < n; col++) {
            if (matrix[0][col] === 0 || matrix[row][0] === 0) {
                matrix[row][col] = 0;
            }
        }
    }

    if (matrix[0][0] === 0) {
        for (let row = 0; row < m; row++) {
            matrix[row][0] = 0;
        }
    }

    if (setZero) {
        for (let col = 0; col < n; col++) {
            matrix[0][col] = 0;
        }
    }

    return matrix;
}

console.log(ZeroMatrix([[1,1,1],[1,0,1],[1,1,1]]));
